src/rest_scripts_execute_POST.ecpp

```diff
@@ -0,0 +1,170 @@
+            {BiosProfile::Admin, "C"}
+    };
+
+    CHECK_USER_PERMISSIONS_OR_DIE(PERMISSIONS);
+
+    // Deserialize request.
+    cxxtools::SerializationInfo in;
+    {
+        std::istringstream iss(request.getBody());
+        cxxtools::JsonDeserializer deserializer(iss);
+        deserializer.deserialize(in);
+    }
+
+    // Grab request variables.
+    std::string name;
+    in.getMember("name").getValue(name);
+    std::string checked_name;
+    check_regex_text_or_die("name", name, checked_name, SCRIPT_FILENAME_REGEX);
+    checked_name = SCRIPT_DIRECTORY + checked_name;
+
+    std::vector<std::string> parameters, env;
+    {
+        std::string array_parsing = "parameters";
+        try
+        {
+            cxxtools::SerializationInfo *siParameters = in.findMember("parameters"), *siEnv = in.findMember("env");
+            if (siParameters) {
+                *siParameters >>= parameters;
+            }
+            array_parsing = "env";
+            if (siEnv) {
+                *siEnv >>= env;
+            }
+        }
+        catch (std::runtime_error &e) {
+            http_die("request-param-bad", array_parsing.c_str(), "Incorrectly formatted", "Array of strings");
+        }
+    }
+
+    std::string timeout;
+    {
+        int timeoutInt = 60;
+        in.getMember("timeout", timeoutInt);
+        timeout = std::to_string(timeoutInt);
+        if (timeoutInt < 1 || timeoutInt > 300) {
+            http_die("request-param-bad", "timeout", timeout.c_str(), "1 <= timeout <= 300");
+        }
+    }
+
+    double execution_time;
+    std::ostringstream output;
+
+    // Set up child's file descriptors.
+    int out_pipe[2], wstatus;
+    posix_spawn_file_actions_t fileActions;
+    pid_t childPid;
+    struct timeval tvStart, tvEnd;
+
+    if (pipe(out_pipe) != 0) {
+        http_die("internal-error", "Error while creating pipe");
+    }
+    posix_spawn_file_actions_init(&fileActions);
+    posix_spawn_file_actions_addclose(&fileActions, STDIN_FILENO);
+    posix_spawn_file_actions_addclose(&fileActions, out_pipe[0]);
+    posix_spawn_file_actions_adddup2(&fileActions, out_pipe[1], STDOUT_FILENO);
+    posix_spawn_file_actions_adddup2(&fileActions, out_pipe[1], STDERR_FILENO);
+    posix_spawn_file_actions_addclose(&fileActions, out_pipe[1]);
+
+    // Build child's argv.
+    std::vector<const char*> argv;
+
+    argv.push_back("/usr/bin/sudo");
+    argv.push_back("/usr/libexec/fty/runas-_bios-script");
+    for (const auto &i : env) {
+        argv.push_back("-e");
+        argv.push_back(i.c_str());
+    }
+    argv.push_back("-t");
+    argv.push_back(timeout.c_str());
+    argv.push_back(checked_name.c_str());
+    for (const auto &i : parameters) {
+        argv.push_back(i.c_str());
+    }
+    argv.push_back(nullptr);
+
+    // Launch child.
+    gettimeofday(&tvStart, nullptr);
+
+    if (posix_spawnp(&childPid, "/usr/bin/sudo", &fileActions, nullptr, (char* const*)&argv[0], environ) != 0) {
```

Comment:
MlmSubprocess class was developped for this goal.
Could you use it in place of your posix_spawnp approach

Reply:
After checking out MlmSubprocess, it does not offer a way to merge STDOUT and STDERR into one stream which is what we want for user scripts. Its implementation must be improved in this regard in order to switch to it.